### Pre-processing for Survival Analysis

In [1]:
import os
import numpy as np
import pandas as pd
import geopandas as gpd
import matplotlib.cm as cm
import matplotlib.pyplot as plt
from requests import get
from urllib.parse import urlparse

In [2]:
df = pd.read_csv("Hampsted HS python.csv", low_memory=False)

In [3]:
df.head()

Unnamed: 0      rates_code rates_authority_id rates_authority  \
0        1540  00000131998018          E09000007          Camden   
1        1541  00000131998018          E09000007          Camden   
2        1542  00000131998018          E09000007          Camden   
3        1543  00000131998018          E09000007          Camden   
4        1544  00000131998018          E09000007          Camden   

                region  location_code          address_no  \
0  Inner London - West      102495063  BST & GND FS 55-56   
1  Inner London - West      102495063  BST & GND FS 55-56   
2  Inner London - West      102495063  BST & GND FS 55-56   
3  Inner London - West      102495063  BST & GND FS 55-56   
4  Inner London - West      102495063  BST & GND FS 55-56   

          address_street address_town  pc__pcs  ... V69  V70        V71 V72  \
0  HAMPSTEAD HIGH STREET       LONDON  NW3 1QH  ...   D  NaN  E05000135 NaN   
1  HAMPSTEAD HIGH STREET       LONDON  NW3 1QH  ...   D  NaN  E05000135 NaN   
2  HAMPSTEAD HIGH STREET       LONDON  NW3 1QH  ...   D  NaN  E05000135 NaN   
3  HAMPSTEAD HIGH STREET       LONDON  NW3 1QH  ...   D  NaN  E05000135 NaN   
4  HAMPSTEAD HIGH STREET       LONDON  NW3 1QH  ...   D  NaN  E05000135 NaN   

          V73  V74  V75   V76 V77                                     geometry  
0  2012-03-19    9  CS1  21.0 NaN  POINT (526376.7504090276 185744.3139116995)  
1  2012-03-19    9  CS1  21.0 NaN  POINT (526376.7504090276 185744.3139116995)  
2  2012-03-19    9  CS1  21.0 NaN  POINT (526376.7504090276 185744.3139116995)  
3  2012-03-19    9  CS1  21.0 NaN  POINT (526376.7504090276 185744.3139116995)  
4  2012-03-19    9  CS1  21.0 NaN  POINT (526376.7504090276 185744.3139116995)  

[5 rows x 104 columns]

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10117 entries, 0 to 10116
Columns: 104 entries, Unnamed: 0 to geometry
dtypes: bool(2), float64(39), int64(12), object(51)
memory usage: 7.9+ MB


In [5]:
df['status_date']

0        2010-04-01
1        2010-04-01
2        2010-04-01
3        2010-04-01
4        2010-04-01
            ...    
10112    2020-03-08
10113    2020-03-08
10114    2020-03-08
10115    2020-11-18
10116    2020-11-18
Name: status_date, Length: 10117, dtype: object

In [6]:
df = df.sort_values(by= ['status_date'])

In [7]:
cols = ['V3', 'V4', 'V5', 'V13', 'V14', 'V15', 'V16', 'V17', 'V19', 'V20', 'V21', 'V26', 'V28', 'V30', 'V36', 'V38', 'V39', 'V48', 'V51', 'V54', 'V55', 'V56', 'V57', 
        'V58', 'V59', 'V60', 'V64', 'V70', 'V72', 'V77']

In [8]:
#deleted those cols columns from the dataset as they only contain nulls

df = df.drop(cols, axis=1)

In [9]:
df['period'].sample(14)

2811    2020-03
7371    2016-03
9038    2015-06
1740    2016-06
5808    2010-06
5538    2019-12
7806    2014-12
1192    2011-06
9865    2020-03
9692    2017-09
42      2020-12
3431    2021-03
4842    2010-12
507     2016-03
Name: period, dtype: object

In [10]:
from datetime import datetime

In [11]:
df['new_date'] = pd.to_datetime(df['status_date'], errors='coerce')

In [12]:
df['new_date']

7210   2015-01-02
5989   2010-01-04
6638   2010-01-04
5450   2010-01-04
5098   2010-01-04
          ...    
6990   2015-09-28
2986   2017-11-28
8629   2009-10-29
489    2009-10-29
258    2015-11-30
Name: new_date, Length: 10117, dtype: datetime64[ns]

In [13]:
nulls = df['new_date'][df['new_date'].notnull()==False]

In [14]:
nulls

Series([], Name: new_date, dtype: datetime64[ns])

In [15]:
df['new_date'].sample(25)

2069   2010-04-01
9891   2017-04-01
4786   2010-04-01
9364   2010-04-01
1205   2010-04-01
5381   2010-04-01
9347   2017-04-01
6063   2010-04-01
3049   2010-04-01
2467   2010-04-01
9168   2010-04-01
2561   2010-04-01
9427   2017-01-04
5727   2010-04-01
8008   2010-04-01
3761   2010-04-01
8509   2010-04-01
3555   2010-04-01
7167   2010-04-01
2800   2010-04-01
9344   2015-01-01
7852   2010-04-01
2693   2011-06-18
4641   2015-06-12
9924   2017-10-27
Name: new_date, dtype: datetime64[ns]

In [16]:
df['status_date'].head(5)

7210    01/02/2015
5989    01/04/2010
6638    01/04/2010
5450    01/04/2010
5098    01/04/2010
Name: status_date, dtype: object

In [17]:
df['new_date'].head(5)

7210   2015-01-02
5989   2010-01-04
6638   2010-01-04
5450   2010-01-04
5098   2010-01-04
Name: new_date, dtype: datetime64[ns]

In [18]:
df = df.sort_values(by= ['new_date'])

In [19]:
df['status_date'].head(5)

8722    2009-10-26
8713    2009-10-26
3231    2009-10-26
8720    2009-10-26
3232    2009-10-26
Name: status_date, dtype: object

In [20]:
df['new_date'].head(5)

8722   2009-10-26
8713   2009-10-26
3231   2009-10-26
8720   2009-10-26
3232   2009-10-26
Name: new_date, dtype: datetime64[ns]

In [21]:
old = df[(df['new_date'] < '2010-04-01')]

In [22]:
old

Unnamed: 0      rates_code rates_authority_id rates_authority  \
8722      306446  99121011200013          E09000007          Camden   
8713      306437  99121011200013          E09000007          Camden   
3231       12911  0006901061004A          E09000007          Camden   
8720      306444  99121011200013          E09000007          Camden   
3232       12912  0006901061004A          E09000007          Camden   
...          ...             ...                ...             ...   
6575       24483  00159000610022          E09000007          Camden   
6557       24465  00159000610022          E09000007          Camden   
6569       24477  00159000610022          E09000007          Camden   
6558       24466  00159000610022          E09000007          Camden   
6561       24469     1.59001E+11          E09000007          Camden   

                   region  location_code          address_no address_street  \
8722  Inner London - West     4017622000  BST & GND FLRS 112   HEATH STREET   
8713  Inner London - West     4017622000  BST & GND FLRS 112   HEATH STREET   
3231  Inner London - West     3942586000    BST & GND FS 106   HEATH STREET   
8720  Inner London - West     4017622000  BST & GND FLRS 112   HEATH STREET   
3232  Inner London - West     3942586000    BST & GND FS 106   HEATH STREET   
...                   ...            ...                 ...            ...   
6575  Inner London - West      100499063      BST & GND FS 6     FLASK WALK   
6557  Inner London - West      100499063      BST & GND FS 6     FLASK WALK   
6569  Inner London - West      100499063      BST & GND FS 6     FLASK WALK   
6558  Inner London - West      100499063      BST & GND FS 6     FLASK WALK   
6561  Inner London - West      100499063      BST & GND FS 6     FLASK WALK   

     address_town  pc__pcs  ... V67  V68 V69        V71         V73  V74 V75  \
8722       LONDON  NW3 1DR  ...   S   1G   D  E05000135  2012-03-19    0  CS   
8713       LONDON  NW3 1DR  ...   S   1G   D  E05000135  2012-03-19    0  CS   
3231       LONDON  NW3 1DR  ...   S   1D   D  E05000135  2012-03-19    1  CS   
8720       LONDON  NW3 1DR  ...   S   1G   D  E05000135  2012-03-19    0  CS   
3232       LONDON  NW3 1DR  ...   S   1D   D  E05000135  2012-03-19    1  CS   
...           ...      ...  ...  ..  ...  ..        ...         ...  ...  ..   
6575       LONDON  NW3 1HE  ...   S   2B   D  E05000135  2012-03-19    1  CS   
6557       LONDON  NW3 1HE  ...   S   2B   D  E05000135  2012-03-19    1  CS   
6569       LONDON  NW3 1HE  ...   S   2B   D  E05000135  2012-03-19    1  CS   
6558       LONDON  NW3 1HE  ...   S   2B   D  E05000135  2012-03-19    1  CS   
6561       LONDON  NW3 1HE  ...   S   2B   D  E05000135  2012-03-19    1  CS   

        V76                                     geometry   new_date  
8722  249.0  POINT (526385.7491817896 186065.3075482332) 2009-10-26  
8713  249.0  POINT (526385.7491817896 186065.3075482332) 2009-10-26  
3231  249.0   POINT (526380.749324891 186014.3085337395) 2009-10-26  
8720  249.0  POINT (526385.7491817896 186065.3075482332) 2009-10-26  
3232  249.0   POINT (526380.749324891 186014.3085337395) 2009-10-26  
...     ...                                          ...        ...  
6575  249.0  POINT (526439.7541909092 185749.3106223808) 2010-03-16  
6557  249.0  POINT (526439.7541909092 185749.3106223808) 2010-03-16  
6569  249.0  POINT (526439.7541909092 185749.3106223808) 2010-03-16  
6558  249.0  POINT (526439.7541909092 185749.3106223808) 2010-03-16  
6561  249.0  POINT (526439.7541909092 185749.3106223808) 2010-03-16  

[546 rows x 75 columns]

546 rows of data where the date is less than our study start date and needs changing to 2010-04-01

In [23]:
df['period'].sample(14)

7624    2013-06
2511    2011-03
3854    2016-12
1934    2020-12
393     2020-09
5777    2013-09
2147    2019-03
2372    2020-06
6074    2010-12
4204    2016-06
2024    2010-06
1370    2011-12
22      2015-12
5174    2016-12
Name: period, dtype: object

In [24]:
df['new_date'].loc[(df['new_date'] < '2010-04-01')] = '2010-04-01'

/opt/conda/envs/sds2020/lib/python3.7/site-packages/pandas/core/indexing.py:670: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  iloc._setitem_with_indexer(indexer, value)


In [25]:
df['new_date'].head(5)

8722    2010-04-01
8713    2010-04-01
3231    2010-04-01
8720    2010-04-01
3232    2010-04-01
Name: new_date, dtype: object

In [26]:
df = df.sort_values(by= ['location_code', 'period'])

In [27]:
df

Unnamed: 0      rates_code rates_authority_id rates_authority  \
7304        25828  00184001B00002          E09000007          Camden   
7305        25829  00184001B00002          E09000007          Camden   
7306        25830  00184001B00002          E09000007          Camden   
7307        25831  00184001B00002          E09000007          Camden   
7308        25832  00184001B00002          E09000007          Camden   
...           ...             ...                ...             ...   
10110      429165  00153007210029          E09000007          Camden   
10111      429166  00153007210029          E09000007          Camden   
10112      429167  00153007210029          E09000007          Camden   
10113      429168  00153007210029          E09000007          Camden   
10114      429169  00153007210029          E09000007          Camden   

                    region  location_code  address_no         address_street  \
7304   Inner London - West       25638063          1B         DOWNSHIRE HILL   
7305   Inner London - West       25638063          1B         DOWNSHIRE HILL   
7306   Inner London - West       25638063          1B         DOWNSHIRE HILL   
7307   Inner London - West       25638063          1B         DOWNSHIRE HILL   
7308   Inner London - West       25638063          1B         DOWNSHIRE HILL   
...                    ...            ...         ...                    ...   
10110  Inner London - West    13174051000  1ST FLR 72  HAMPSTEAD HIGH STREET   
10111  Inner London - West    13174051000  1ST FLR 72  HAMPSTEAD HIGH STREET   
10112  Inner London - West    13174051000  1ST FLR 72  HAMPSTEAD HIGH STREET   
10113  Inner London - West    13174051000  1ST FLR 72  HAMPSTEAD HIGH STREET   
10114  Inner London - West    13174051000  1ST FLR 72  HAMPSTEAD HIGH STREET   

      address_town  pc__pcs  ...  V67  V68 V69        V71         V73  V74  \
7304        LONDON  NW3 1NR  ...    S   2G   D  E05000135  2012-03-19    0   
7305        LONDON  NW3 1NR  ...    S   2G   D  E05000135  2012-03-19    0   
7306        LONDON  NW3 1NR  ...    S   2G   D  E05000135  2012-03-19    0   
7307        LONDON  NW3 1NR  ...    S   2G   D  E05000135  2012-03-19    0   
7308        LONDON  NW3 1NR  ...    S   2G   D  E05000135  2012-03-19    0   
...            ...      ...  ...  ...  ...  ..        ...         ...  ...   
10110       LONDON  NW3 1QP  ...  NaN  NaN   C  E05000135         NaN    0   
10111       LONDON  NW3 1QP  ...  NaN  NaN   C  E05000135         NaN    0   
10112       LONDON  NW3 1QP  ...  NaN  NaN   C  E05000135         NaN    0   
10113       LONDON  NW3 1QP  ...  NaN  NaN   C  E05000135         NaN    0   
10114       LONDON  NW3 1QP  ...  NaN  NaN   C  E05000135         NaN    0   

      V75    V76                                     geometry  \
7304   MH  437.0  POINT (526838.7620297795 185589.3079796332)   
7305   MH  437.0  POINT (526838.7620297795 185589.3079796332)   
7306   MH  437.0  POINT (526838.7620297795 185589.3079796332)   
7307   MH  437.0  POINT (526838.7620297795 185589.3079796332)   
7308   MH  437.0  POINT (526838.7620297795 185589.3079796332)   
...    ..    ...                                          ...   
10110  CO  203.0  POINT (526450.7546235088 185680.3055356044)   
10111  CO  203.0  POINT (526450.7546235088 185680.3055356044)   
10112  CO  203.0  POINT (526450.7546235088 185680.3055356044)   
10113  CO  203.0  POINT (526450.7546235088 185680.3055356044)   
10114  CO  203.0  POINT (526450.7546235088 185680.3055356044)   

                  new_date  
7304   2010-04-01 00:00:00  
7305   2010-04-01 00:00:00  
7306   2010-04-01 00:00:00  
7307   2010-04-01 00:00:00  
7308   2010-04-01 00:00:00  
...                    ...  
10110  2020-03-08 00:00:00  
10111  2020-03-08 00:00:00  
10112  2020-03-08 00:00:00  
10113  2020-03-08 00:00:00  
10114  2020-03-08 00:00:00  

[10117 rows x 75 columns]

### CHECK FOR VACANT STORES

In [28]:
vacant = df[(df['status'] == False)]

In [29]:
vacant

Unnamed: 0      rates_code rates_authority_id rates_authority  \
7304        25828  00184001B00002          E09000007          Camden   
7305        25829  00184001B00002          E09000007          Camden   
7306        25830  00184001B00002          E09000007          Camden   
7307        25831  00184001B00002          E09000007          Camden   
7308        25832  00184001B00002          E09000007          Camden   
...           ...             ...                ...             ...   
10093      419528  0029500482000A          E09000007          Camden   
10094      419529  0029500482000A          E09000007          Camden   
10095      419530  0029500482000A          E09000007          Camden   
10096      419531  0029500482000A          E09000007          Camden   
10097      419532  0029500482000A          E09000007          Camden   

                    region  location_code address_no  address_street  \
7304   Inner London - West       25638063         1B  DOWNSHIRE HILL   
7305   Inner London - West       25638063         1B  DOWNSHIRE HILL   
7306   Inner London - West       25638063         1B  DOWNSHIRE HILL   
7307   Inner London - West       25638063         1B  DOWNSHIRE HILL   
7308   Inner London - West       25638063         1B  DOWNSHIRE HILL   
...                    ...            ...        ...             ...   
10093  Inner London - West    11765732000         48    ROSSLYN HILL   
10094  Inner London - West    11765732000         48    ROSSLYN HILL   
10095  Inner London - West    11765732000         48    ROSSLYN HILL   
10096  Inner London - West    11765732000         48    ROSSLYN HILL   
10097  Inner London - West    11765732000         48    ROSSLYN HILL   

      address_town  pc__pcs  ... V67  V68 V69        V71         V73  V74 V75  \
7304        LONDON  NW3 1NR  ...   S   2G   D  E05000135  2012-03-19    0  MH   
7305        LONDON  NW3 1NR  ...   S   2G   D  E05000135  2012-03-19    0  MH   
7306        LONDON  NW3 1NR  ...   S   2G   D  E05000135  2012-03-19    0  MH   
7307        LONDON  NW3 1NR  ...   S   2G   D  E05000135  2012-03-19    0  MH   
7308        LONDON  NW3 1NR  ...   S   2G   D  E05000135  2012-03-19    0  MH   
...            ...      ...  ...  ..  ...  ..        ...         ...  ...  ..   
10093       LONDON  NW3 1NH  ...   S   1J   D  E05000135  2012-03-19    1  CR   
10094       LONDON  NW3 1NH  ...   S   1J   D  E05000135  2012-03-19    1  CR   
10095       LONDON  NW3 1NH  ...   S   1J   D  E05000135  2012-03-19    1  CR   
10096       LONDON  NW3 1NH  ...   S   1J   D  E05000135  2012-03-19    1  CR   
10097       LONDON  NW3 1NH  ...   S   1J   D  E05000135  2012-03-19    1  CR   

         V76                                     geometry             new_date  
7304   437.0  POINT (526838.7620297795 185589.3079796332)  2010-04-01 00:00:00  
7305   437.0  POINT (526838.7620297795 185589.3079796332)  2010-04-01 00:00:00  
7306   437.0  POINT (526838.7620297795 185589.3079796332)  2010-04-01 00:00:00  
7307   437.0  POINT (526838.7620297795 185589.3079796332)  2010-04-01 00:00:00  
7308   437.0  POINT (526838.7620297795 185589.3079796332)  2010-04-01 00:00:00  
...      ...                                          ...                  ...  
10093  234.0  POINT (526753.7573694737 185603.3076923557)  2019-07-16 00:00:00  
10094  234.0  POINT (526753.7573694737 185603.3076923557)  2019-07-16 00:00:00  
10095  234.0  POINT (526753.7573694737 185603.3076923557)  2019-07-16 00:00:00  
10096  234.0  POINT (526753.7573694737 185603.3076923557)  2019-07-16 00:00:00  
10097  234.0  POINT (526753.7573694737 185603.3076923557)  2019-07-16 00:00:00  

[610 rows x 75 columns]

In [30]:
# 610 vacant stores but need to establish if any were vacant prior to study period 

## LEFT CENSOR 

In [31]:
# do i need to make the location codes keys?

In [32]:
remove = vacant[(vacant['status_date'] == '2010-04-01')]

In [33]:
remove

Unnamed: 0      rates_code rates_authority_id rates_authority  \
7304       25828  00184001B00002          E09000007          Camden   
7305       25829  00184001B00002          E09000007          Camden   
7306       25830  00184001B00002          E09000007          Camden   
7307       25831  00184001B00002          E09000007          Camden   
7308       25832  00184001B00002          E09000007          Camden   
...          ...             ...                ...             ...   
3515       21027  00148000100036          E09000007          Camden   
3516       21028  00148000100036          E09000007          Camden   
3517       21029  00148000100036          E09000007          Camden   
3518       21030  00148000100036          E09000007          Camden   
3519       21031  00148000100036          E09000007          Camden   

                   region  location_code address_no  address_street  \
7304  Inner London - West       25638063         1B  DOWNSHIRE HILL   
7305  Inner London - West       25638063         1B  DOWNSHIRE HILL   
7306  Inner London - West       25638063         1B  DOWNSHIRE HILL   
7307  Inner London - West       25638063         1B  DOWNSHIRE HILL   
7308  Inner London - West       25638063         1B  DOWNSHIRE HILL   
...                   ...            ...        ...             ...   
3515  Inner London - West     5729723000  3RD FLR 1      HOLLY HILL   
3516  Inner London - West     5729723000  3RD FLR 1      HOLLY HILL   
3517  Inner London - West     5729723000  3RD FLR 1      HOLLY HILL   
3518  Inner London - West     5729723000  3RD FLR 1      HOLLY HILL   
3519  Inner London - West     5729723000  3RD FLR 1      HOLLY HILL   

     address_town  pc__pcs  ...  V67  V68 V69        V71         V73  V74 V75  \
7304       LONDON  NW3 1NR  ...    S   2G   D  E05000135  2012-03-19    0  MH   
7305       LONDON  NW3 1NR  ...    S   2G   D  E05000135  2012-03-19    0  MH   
7306       LONDON  NW3 1NR  ...    S   2G   D  E05000135  2012-03-19    0  MH   
7307       LONDON  NW3 1NR  ...    S   2G   D  E05000135  2012-03-19    0  MH   
7308       LONDON  NW3 1NR  ...    S   2G   D  E05000135  2012-03-19    0  MH   
...           ...      ...  ...  ...  ...  ..        ...         ...  ...  ..   
3515       LONDON  NW3 6UB  ...  NaN  NaN   C  E05000133         NaN    0  CO   
3516       LONDON  NW3 6UB  ...  NaN  NaN   C  E05000133         NaN    0  CO   
3517       LONDON  NW3 6UB  ...  NaN  NaN   C  E05000133         NaN    0  CO   
3518       LONDON  NW3 6UB  ...  NaN  NaN   C  E05000133         NaN    0  CO   
3519       LONDON  NW3 6UB  ...  NaN  NaN   C  E05000133         NaN    0  CO   

        V76                                     geometry             new_date  
7304  437.0  POINT (526838.7620297795 185589.3079796332)  2010-04-01 00:00:00  
7305  437.0  POINT (526838.7620297795 185589.3079796332)  2010-04-01 00:00:00  
7306  437.0  POINT (526838.7620297795 185589.3079796332)  2010-04-01 00:00:00  
7307  437.0  POINT (526838.7620297795 185589.3079796332)  2010-04-01 00:00:00  
7308  437.0  POINT (526838.7620297795 185589.3079796332)  2010-04-01 00:00:00  
...     ...                                          ...                  ...  
3515  203.0    POINT (526353.754357001 185775.309671689)  2010-04-01 00:00:00  
3516  203.0    POINT (526353.754357001 185775.309671689)  2010-04-01 00:00:00  
3517  203.0    POINT (526353.754357001 185775.309671689)  2010-04-01 00:00:00  
3518  203.0    POINT (526353.754357001 185775.309671689)  2010-04-01 00:00:00  
3519  203.0    POINT (526353.754357001 185775.309671689)  2010-04-01 00:00:00  

[125 rows x 75 columns]

In [34]:
df['locaton_code'] = df['location_code'].astype("category")

In [35]:
df['location_code'].unique()

array([   25638063,    32905063,    32907063,    32912063,    32945063,
          32948063,    41376063,    41380063,    41404063,    41419063,
          41422063,    41424063,    41429063,    41440063,    41452063,
          41465063,    41468063,    41472063,    41481063,    44969063,
          44977063,    44984063,    44990063,    44992063,    45002063,
          45010063,    45060063,    45067063,    45075063,    45077063,
          45082063,    45099063,    45111063,    45113063,    45118063,
          45119063,    46955063,    47002063,    67617063,    67618063,
          74478063,    95061063,    95068063,    95134063,    95135063,
          98036063,    98039063,    98040063,    98041063,    98043063,
          98048063,    98049063,    98050063,    98051063,    98052063,
          98053063,    98054063,    98056063,    98057063,   100496063,
         100498063,   100499063,   100501063,   100790063,   102339063,
         102342063,   102345063,   102353063,   102354063,   102

In [36]:
df['location_code'].value_counts()

220547063      44
2482984000     44
100790063      44
3942586000     44
176189063      44
               ..
11929598000    13
12229537000     7
13174048000     5
13174051000     5
13026783000     2
Name: location_code, Length: 261, dtype: int64

In [37]:
 # 261 unique locations in Hampstead HighStreet

In [38]:
locations = df['location_code'].unique()

In [39]:
df.loc[df['name'].isnull(), 'name'] = df['location_code']

In [40]:
df

Unnamed: 0      rates_code rates_authority_id rates_authority  \
7304        25828  00184001B00002          E09000007          Camden   
7305        25829  00184001B00002          E09000007          Camden   
7306        25830  00184001B00002          E09000007          Camden   
7307        25831  00184001B00002          E09000007          Camden   
7308        25832  00184001B00002          E09000007          Camden   
...           ...             ...                ...             ...   
10110      429165  00153007210029          E09000007          Camden   
10111      429166  00153007210029          E09000007          Camden   
10112      429167  00153007210029          E09000007          Camden   
10113      429168  00153007210029          E09000007          Camden   
10114      429169  00153007210029          E09000007          Camden   

                    region  location_code  address_no         address_street  \
7304   Inner London - West       25638063          1B         DOWNSHIRE HILL   
7305   Inner London - West       25638063          1B         DOWNSHIRE HILL   
7306   Inner London - West       25638063          1B         DOWNSHIRE HILL   
7307   Inner London - West       25638063          1B         DOWNSHIRE HILL   
7308   Inner London - West       25638063          1B         DOWNSHIRE HILL   
...                    ...            ...         ...                    ...   
10110  Inner London - West    13174051000  1ST FLR 72  HAMPSTEAD HIGH STREET   
10111  Inner London - West    13174051000  1ST FLR 72  HAMPSTEAD HIGH STREET   
10112  Inner London - West    13174051000  1ST FLR 72  HAMPSTEAD HIGH STREET   
10113  Inner London - West    13174051000  1ST FLR 72  HAMPSTEAD HIGH STREET   
10114  Inner London - West    13174051000  1ST FLR 72  HAMPSTEAD HIGH STREET   

      address_town  pc__pcs  ...  V68  V69        V71         V73  V74  V75  \
7304        LONDON  NW3 1NR  ...   2G    D  E05000135  2012-03-19    0   MH   
7305        LONDON  NW3 1NR  ...   2G    D  E05000135  2012-03-19    0   MH   
7306        LONDON  NW3 1NR  ...   2G    D  E05000135  2012-03-19    0   MH   
7307        LONDON  NW3 1NR  ...   2G    D  E05000135  2012-03-19    0   MH   
7308        LONDON  NW3 1NR  ...   2G    D  E05000135  2012-03-19    0   MH   
...            ...      ...  ...  ...  ...        ...         ...  ...  ...   
10110       LONDON  NW3 1QP  ...  NaN    C  E05000135         NaN    0   CO   
10111       LONDON  NW3 1QP  ...  NaN    C  E05000135         NaN    0   CO   
10112       LONDON  NW3 1QP  ...  NaN    C  E05000135         NaN    0   CO   
10113       LONDON  NW3 1QP  ...  NaN    C  E05000135         NaN    0   CO   
10114       LONDON  NW3 1QP  ...  NaN    C  E05000135         NaN    0   CO   

         V76                                     geometry  \
7304   437.0  POINT (526838.7620297795 185589.3079796332)   
7305   437.0  POINT (526838.7620297795 185589.3079796332)   
7306   437.0  POINT (526838.7620297795 185589.3079796332)   
7307   437.0  POINT (526838.7620297795 185589.3079796332)   
7308   437.0  POINT (526838.7620297795 185589.3079796332)   
...      ...                                          ...   
10110  203.0  POINT (526450.7546235088 185680.3055356044)   
10111  203.0  POINT (526450.7546235088 185680.3055356044)   
10112  203.0  POINT (526450.7546235088 185680.3055356044)   
10113  203.0  POINT (526450.7546235088 185680.3055356044)   
10114  203.0  POINT (526450.7546235088 185680.3055356044)   

                  new_date  locaton_code  
7304   2010-04-01 00:00:00      25638063  
7305   2010-04-01 00:00:00      25638063  
7306   2010-04-01 00:00:00      25638063  
7307   2010-04-01 00:00:00      25638063  
7308   2010-04-01 00:00:00      25638063  
...                    ...           ...  
10110  2020-03-08 00:00:00   13174051000  
10111  2020-03-08 00:00:00   13174051000  
10112  2020-03-08 00:00:00   13174051000  
10113  2020-03-08 00:00:00   13174051000  
10114  2020-03-08 00:00:00

In [41]:
df['status_date'] = pd.to_datetime(df['status_date'], errors='coerce')

In [42]:
df['status_date'].iloc[df['status_date'] < '2010-04-01'] = '04/01/2010'

/opt/conda/envs/sds2020/lib/python3.7/site-packages/pandas/core/indexing.py:670: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  iloc._setitem_with_indexer(indexer, value)


In [43]:
df['status_date'] = pd.to_datetime(df['status_date'], errors='coerce')

In [44]:
df.sample(15)

Unnamed: 0      rates_code rates_authority_id rates_authority  \
1100       10780  00069003000001          E09000007          Camden   
9064      338850  0006900821001B          E09000007          Camden   
5377       22889  00153006200004          E09000007          Camden   
483        10163  00069000510022          E09000007          Camden   
1175       10855  00069003100002          E09000007          Camden   
8714      306438  99121011200013          E09000007          Camden   
578        10258  0006900090001A          E09000007          Camden   
9403      366946  000690024A0009          E09000007          Camden   
9389      366932  000690024A0009          E09000007          Camden   
1767       11447  00069005120002          E09000007          Camden   
8977      332726  00184000220001          E09000007          Camden   
3545       21057  0014800031002A          E09000007          Camden   
3160       12840  00069010410014          E09000007          Camden   
6576       24484  00159000610022          E09000007          Camden   
1482       11162  00069003910016          E09000007          Camden   

                   region  location_code                       address_no  \
1100  Inner London - West      112447063                BST & GND F(S) 30   
9064  Inner London - West      272838000                  BST & GND FS 82   
5377  Inner London - West       41404063                               62   
483   Inner London - West      104876063                          1ST F 5   
1175  Inner London - West      104898063                               31   
8714  Inner London - West     4017622000               BST & GND FLRS 112   
578   Inner London - West      112596063                   BST & GND FS 9   
9403  Inner London - West      494438000                              24A   
9389  Inner London - West      494438000                              24A   
1767  Inner London - West      104908063  EXC 1ST & 2ND FS OF 53 AT 51-53   
8977  Inner London - West     8408395000                          2A & 3A   
3545  Inner London - West      112626063                 BST & GRD FLRS 3   
3160  Inner London - West      105070063                 BST & GND FS 104   
6576  Inner London - West      100499063                   BST & GND FS 6   
1482  Inner London - West      104899063                  BST & GND FS 39   

             address_street address_town  pc__pcs  ...  V68  V69        V71  \
1100           HEATH STREET       LONDON  NW3 6TE  ...   1L    D  E05000135   
9064           HEATH STREET       LONDON  NW3 1DN  ...   1L    D  E05000135   
5377  HAMPSTEAD HIGH STREET       LONDON  NW3 1QH  ...  NaN    L  E05000135   
483            HEATH STREET       LONDON  NW3 6TP  ...  NaN    L  E05000133   
1175           HEATH STREET       LONDON  NW3 6TR  ...   1U    D  E05000133   
8714           HEATH STREET       LONDON  NW3 1DR  ...   1G    D  E05000135   
578            HEATH STREET       LONDON  NW3 6TP  ...   1D    D  E05000133   
9403           HEATH STREET       LONDON  NW3 6TE  ...   2B    D  E05000135   
9389           HEATH STREET       LONDON  NW3 6TE  ...   2B    D  E05000135   
1767           HEATH STREET       LONDON  NW3 6UG  ...   1A    D  E05000135   
8977         DOWNSHIRE HILL       LONDON  NW3 1NR  ...  NaN    L  E05000135   
3545             HOLLY HILL       LONDON  NW3 6QN  ...   1G    D  E05000133   
3160           HEATH STREET       LONDON  NW3 1DR  ...   1B    D  E05000135   
6576             FLASK WALK       LONDON  NW3 1HE  ...   2B    D  E05000135   
1482           HEATH STREET       LONDON  NW3 6UA  ...   1A    D  E05000133   

             V73  V74  V75    V76  \
1100  2012-03-19    1  CS3  417.0   
9064  2012-03-19    1   CS  249.0   
5377         NaN    1   CS  249.0   
483          NaN    0   CO  203.0   
1175  2012-03-19    1   CS  249.0   
8714  2012-03-19    0   CS  249.0   
578   2012-03-19    2   CO  203.0   
9403  2012-03-19    0   LX  416.0   
9389  2012-03-19    0   LX  416.0   
1

### change period to a date format eg. 2021-03 = 01/04/2021

In [45]:
df['period2'] = pd.to_datetime(df['period'])

In [46]:
df['period2'] = df['period2'] + pd.DateOffset(months=1)

In [47]:
df['duration'] = df['period2'] - df['status_date']

In [48]:
df['duration'] = df['duration'].dt.days

In [49]:
df

Unnamed: 0      rates_code rates_authority_id rates_authority  \
7304        25828  00184001B00002          E09000007          Camden   
7305        25829  00184001B00002          E09000007          Camden   
7306        25830  00184001B00002          E09000007          Camden   
7307        25831  00184001B00002          E09000007          Camden   
7308        25832  00184001B00002          E09000007          Camden   
...           ...             ...                ...             ...   
10110      429165  00153007210029          E09000007          Camden   
10111      429166  00153007210029          E09000007          Camden   
10112      429167  00153007210029          E09000007          Camden   
10113      429168  00153007210029          E09000007          Camden   
10114      429169  00153007210029          E09000007          Camden   

                    region  location_code  address_no         address_street  \
7304   Inner London - West       25638063          1B         DOWNSHIRE HILL   
7305   Inner London - West       25638063          1B         DOWNSHIRE HILL   
7306   Inner London - West       25638063          1B         DOWNSHIRE HILL   
7307   Inner London - West       25638063          1B         DOWNSHIRE HILL   
7308   Inner London - West       25638063          1B         DOWNSHIRE HILL   
...                    ...            ...         ...                    ...   
10110  Inner London - West    13174051000  1ST FLR 72  HAMPSTEAD HIGH STREET   
10111  Inner London - West    13174051000  1ST FLR 72  HAMPSTEAD HIGH STREET   
10112  Inner London - West    13174051000  1ST FLR 72  HAMPSTEAD HIGH STREET   
10113  Inner London - West    13174051000  1ST FLR 72  HAMPSTEAD HIGH STREET   
10114  Inner London - West    13174051000  1ST FLR 72  HAMPSTEAD HIGH STREET   

      address_town  pc__pcs  ...        V71         V73 V74 V75    V76  \
7304        LONDON  NW3 1NR  ...  E05000135  2012-03-19   0  MH  437.0   
7305        LONDON  NW3 1NR  ...  E05000135  2012-03-19   0  MH  437.0   
7306        LONDON  NW3 1NR  ...  E05000135  2012-03-19   0  MH  437.0   
7307        LONDON  NW3 1NR  ...  E05000135  2012-03-19   0  MH  437.0   
7308        LONDON  NW3 1NR  ...  E05000135  2012-03-19   0  MH  437.0   
...            ...      ...  ...        ...         ...  ..  ..    ...   
10110       LONDON  NW3 1QP  ...  E05000135         NaN   0  CO  203.0   
10111       LONDON  NW3 1QP  ...  E05000135         NaN   0  CO  203.0   
10112       LONDON  NW3 1QP  ...  E05000135         NaN   0  CO  203.0   
10113       LONDON  NW3 1QP  ...  E05000135         NaN   0  CO  203.0   
10114       LONDON  NW3 1QP  ...  E05000135         NaN   0  CO  203.0   

                                          geometry             new_date  \
7304   POINT (526838.7620297795 185589.3079796332)  2010-04-01 00:00:00   
7305   POINT (526838.7620297795 185589.3079796332)  2010-04-01 00:00:00   
7306   POINT (526838.7620297795 185589.3079796332)  2010-04-01 00:00:00   
7307   POINT (526838.7620297795 185589.3079796332)  2010-04-01 00:00:00   
7308   POINT (526838.7620297795 185589.3079796332)  2010-04-01 00:00:00   
...                                            ...                  ...   
10110  POINT (526450.7546235088 185680.3055356044)  2020-03-08 00:00:00   
10111  POINT (526450.7546235088 185680.3055356044)  2020-03-08 00:00:00   
10112  POINT (526450.7546235088 185680.3055356044)  2020-03-08 00:00:00   
10113  POINT (526450.7546235088 185680.3055356044)  2020-03-08 00:00:00   
10114  POINT (526450.7546235088 185680.3055356044)  2020-03-08 00:00:00   

       locaton_code    period2  duration  
7304       25638063 2010-07-01        91  
7305       25638063 2010-10-01       183  
7306       25638063 2011-01-01       275  
7307       25638063 2011-04-01       365  
7308       25638063 2011-07-01       456  
...             ...        ...       ...  
10110   13174051000 2020-04-01        24  
10111   13174051000 2020-07-01       115  
1011

In [50]:
df['study-start'] = '04/01/2010'

In [51]:
df['study-start'] = pd.to_datetime(df['study-start'], errors='coerce')

In [52]:
df['time-of-entry'] = df['status_date'] -  df['study-start']

In [53]:
df['time-at-vacant'] = 0

In [54]:
df['vacant'] = 0

In [55]:
df

Unnamed: 0      rates_code rates_authority_id rates_authority  \
7304        25828  00184001B00002          E09000007          Camden   
7305        25829  00184001B00002          E09000007          Camden   
7306        25830  00184001B00002          E09000007          Camden   
7307        25831  00184001B00002          E09000007          Camden   
7308        25832  00184001B00002          E09000007          Camden   
...           ...             ...                ...             ...   
10110      429165  00153007210029          E09000007          Camden   
10111      429166  00153007210029          E09000007          Camden   
10112      429167  00153007210029          E09000007          Camden   
10113      429168  00153007210029          E09000007          Camden   
10114      429169  00153007210029          E09000007          Camden   

                    region  location_code  address_no         address_street  \
7304   Inner London - West       25638063          1B         DOWNSHIRE HILL   
7305   Inner London - West       25638063          1B         DOWNSHIRE HILL   
7306   Inner London - West       25638063          1B         DOWNSHIRE HILL   
7307   Inner London - West       25638063          1B         DOWNSHIRE HILL   
7308   Inner London - West       25638063          1B         DOWNSHIRE HILL   
...                    ...            ...         ...                    ...   
10110  Inner London - West    13174051000  1ST FLR 72  HAMPSTEAD HIGH STREET   
10111  Inner London - West    13174051000  1ST FLR 72  HAMPSTEAD HIGH STREET   
10112  Inner London - West    13174051000  1ST FLR 72  HAMPSTEAD HIGH STREET   
10113  Inner London - West    13174051000  1ST FLR 72  HAMPSTEAD HIGH STREET   
10114  Inner London - West    13174051000  1ST FLR 72  HAMPSTEAD HIGH STREET   

      address_town  pc__pcs  ...    V76  \
7304        LONDON  NW3 1NR  ...  437.0   
7305        LONDON  NW3 1NR  ...  437.0   
7306        LONDON  NW3 1NR  ...  437.0   
7307        LONDON  NW3 1NR  ...  437.0   
7308        LONDON  NW3 1NR  ...  437.0   
...            ...      ...  ...    ...   
10110       LONDON  NW3 1QP  ...  203.0   
10111       LONDON  NW3 1QP  ...  203.0   
10112       LONDON  NW3 1QP  ...  203.0   
10113       LONDON  NW3 1QP  ...  203.0   
10114       LONDON  NW3 1QP  ...  203.0   

                                          geometry             new_date  \
7304   POINT (526838.7620297795 185589.3079796332)  2010-04-01 00:00:00   
7305   POINT (526838.7620297795 185589.3079796332)  2010-04-01 00:00:00   
7306   POINT (526838.7620297795 185589.3079796332)  2010-04-01 00:00:00   
7307   POINT (526838.7620297795 185589.3079796332)  2010-04-01 00:00:00   
7308   POINT (526838.7620297795 185589.3079796332)  2010-04-01 00:00:00   
...                                            ...                  ...   
10110  POINT (526450.7546235088 185680.3055356044)  2020-03-08 00:00:00   
10111  POINT (526450.7546235088 185680.3055356044)  2020-03-08 00:00:00   
10112  POINT (526450.7546235088 185680.3055356044)  2020-03-08 00:00:00   
10113  POINT (526450.7546235088 185680.3055356044)  2020-03-08 00:00:00   
10114  POINT (526450.7546235088 185680.3055356044)  2020-03-08 00:00:00   

      locaton_code    period2  duration study-start  time-of-entry  \
7304      25638063 2010-07-01        91  2010-04-01         0 days   
7305      25638063 2010-10-01       183  2010-04-01         0 days   
7306      25638063 2011-01-01       275  2010-04-01         0 days   
7307      25638063 2011-04-01       365  2010-04-01         0 days   
7308      25638063 2011-07-01       456  2010-04-01         0 days   
...            ...        ...       ...         ...            ...   
10110  13174051000 2020-04-01        24  2010-04-01      3629 days   
10111  13174051000 2020-07-01       115  2010-04-01      3629 days   
10112  13174051000 2020-10-01       207  2010-04-01      3629 days   
10113  13174051000 2021-01-01       299  2010-04-01      3629 d

In [56]:
drop = df[(df['status'] == False) & (df['status_date'] == '2010-04-01')]

In [57]:
drop.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 134 entries, 7304 to 3519
Data columns (total 82 columns):
 #   Column                Non-Null Count  Dtype          
---  ------                --------------  -----          
 0   Unnamed: 0            134 non-null    int64          
 1   rates_code            134 non-null    object         
 2   rates_authority_id    134 non-null    object         
 3   rates_authority       134 non-null    object         
 4   region                134 non-null    object         
 5   location_code         134 non-null    int64          
 6   address_no            134 non-null    object         
 7   address_street        134 non-null    object         
 8   address_town          134 non-null    object         
 9   pc__pcs               134 non-null    object         
 10  use_category          134 non-null    object         
 11  scat_code             134 non-null    int64          
 12  sub_category          134 non-null    object         
 13  c

In [58]:
# 134 rows will be dropped due to being vacant before the study started

In [59]:
df = df.drop(df[(df['status'] == False) & (df['status_date'] == '2010-04-01')].index)

In [60]:
df

Unnamed: 0      rates_code rates_authority_id rates_authority  \
7309        25833  00184001B00002          E09000007          Camden   
7310        25834  00184001B00002          E09000007          Camden   
7311        25835  00184001B00002          E09000007          Camden   
7312        25836  00184001B00002          E09000007          Camden   
7313        25837  00184001B00002          E09000007          Camden   
...           ...             ...                ...             ...   
10110      429165  00153007210029          E09000007          Camden   
10111      429166  00153007210029          E09000007          Camden   
10112      429167  00153007210029          E09000007          Camden   
10113      429168  00153007210029          E09000007          Camden   
10114      429169  00153007210029          E09000007          Camden   

                    region  location_code  address_no         address_street  \
7309   Inner London - West       25638063          1B         DOWNSHIRE HILL   
7310   Inner London - West       25638063          1B         DOWNSHIRE HILL   
7311   Inner London - West       25638063          1B         DOWNSHIRE HILL   
7312   Inner London - West       25638063          1B         DOWNSHIRE HILL   
7313   Inner London - West       25638063          1B         DOWNSHIRE HILL   
...                    ...            ...         ...                    ...   
10110  Inner London - West    13174051000  1ST FLR 72  HAMPSTEAD HIGH STREET   
10111  Inner London - West    13174051000  1ST FLR 72  HAMPSTEAD HIGH STREET   
10112  Inner London - West    13174051000  1ST FLR 72  HAMPSTEAD HIGH STREET   
10113  Inner London - West    13174051000  1ST FLR 72  HAMPSTEAD HIGH STREET   
10114  Inner London - West    13174051000  1ST FLR 72  HAMPSTEAD HIGH STREET   

      address_town  pc__pcs  ...    V76  \
7309        LONDON  NW3 1NR  ...  437.0   
7310        LONDON  NW3 1NR  ...  437.0   
7311        LONDON  NW3 1NR  ...  437.0   
7312        LONDON  NW3 1NR  ...  437.0   
7313        LONDON  NW3 1NR  ...  437.0   
...            ...      ...  ...    ...   
10110       LONDON  NW3 1QP  ...  203.0   
10111       LONDON  NW3 1QP  ...  203.0   
10112       LONDON  NW3 1QP  ...  203.0   
10113       LONDON  NW3 1QP  ...  203.0   
10114       LONDON  NW3 1QP  ...  203.0   

                                          geometry             new_date  \
7309   POINT (526838.7620297795 185589.3079796332)  2011-08-08 00:00:00   
7310   POINT (526838.7620297795 185589.3079796332)  2011-08-08 00:00:00   
7311   POINT (526838.7620297795 185589.3079796332)  2011-08-08 00:00:00   
7312   POINT (526838.7620297795 185589.3079796332)  2011-08-08 00:00:00   
7313   POINT (526838.7620297795 185589.3079796332)  2011-08-08 00:00:00   
...                                            ...                  ...   
10110  POINT (526450.7546235088 185680.3055356044)  2020-03-08 00:00:00   
10111  POINT (526450.7546235088 185680.3055356044)  2020-03-08 00:00:00   
10112  POINT (526450.7546235088 185680.3055356044)  2020-03-08 00:00:00   
10113  POINT (526450.7546235088 185680.3055356044)  2020-03-08 00:00:00   
10114  POINT (526450.7546235088 185680.3055356044)  2020-03-08 00:00:00   

      locaton_code    period2  duration study-start  time-of-entry  \
7309      25638063 2011-10-01        54  2010-04-01       494 days   
7310      25638063 2012-01-01       146  2010-04-01       494 days   
7311      25638063 2012-04-01       237  2010-04-01       494 days   
7312      25638063 2012-07-01       328  2010-04-01       494 days   
7313      25638063 2012-10-01       420  2010-04-01       494 days   
...            ...        ...       ...         ...            ...   
10110  13174051000 2020-04-01        24  2010-04-01      3629 days   
10111  13174051000 2020-07-01       115  2010-04-01      3629 days   
10112  13174051000 2020-10-01       207  2010-04-01      3629 days   
10113  13174051000 2021-01-01       299  2010-04-01      3629 d

In [61]:
# 10117 - 134 = 9983 rows left

In [62]:
df['location_code'] = df['location_code'].astype("category")

In [63]:
locations = list(df['location_code'].unique())

In [64]:
df = df.set_index('location_code')

In [65]:
df['survival'] = 0

In [66]:
df['survival'] = df['duration']

In [67]:
df['entry'] = df['time-of-entry'].dt.days

In [68]:
df['time-at-vacant'] = df['survival'] + df['entry']

In [69]:
df['stat'] = df['status'].astype(int)

In [70]:
df["StatusChange"] =  df['stat'].shift() != df['stat']

In [71]:
df["StatusChange"].iloc[0] = False

/opt/conda/envs/sds2020/lib/python3.7/site-packages/pandas/core/indexing.py:670: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  iloc._setitem_with_indexer(indexer, value)


In [72]:
df["TenantChange"] =  df['name'].shift(-1) != df['name'] 

In [73]:
df["TenantChange"].iloc[0] = False

In [74]:
subset = df[(df['StatusChange'] == False)  & (df['TenantChange'] == True) & (df['status'] == True)]

In [75]:
subset

Unnamed: 0       rates_code rates_authority_id rates_authority  \
location_code                                                                   
25638063            25871   00184001B00002          E09000007          Camden   
32905063            24639   00159001009992          E09000007          Camden   
32907063            24701   00159001200006          E09000007          Camden   
32907063            24727   00159001200006          E09000007          Camden   
32912063            24389   00159000410016          E09000007          Camden   
...                   ...              ...                ...             ...   
12358523000        399173   00153003210013          E09000007          Camden   
12514325000        397925     TEMP33949894          E09000007          Camden   
13026783000        429519  000760029300001          E09000007          Camden   
13174048000        429164   00153007210028          E09000007          Camden   
13174051000        429169   00153007210029          E09000007          Camden   

                            region                      address_no  \
location_code                                                        
25638063       Inner London - West                              1B   
32905063       Inner London - West                              10   
32907063       Inner London - West                              12   
32907063       Inner London - West                              12   
32912063       Inner London - West                               4   
...                            ...                             ...   
12358523000    Inner London - West         1ST FLR (SOUTH WEST) 32   
12514325000    Inner London - West  GND FLR 7-9 & 1ST FLR LEFT 7-9   
13026783000    Inner London - West                     APARTMENT 8   
13174048000    Inner London - West                      GND FLR 72   
13174051000    Inner London - West                      1ST FLR 72   

                      address_street address_town  pc__pcs  \
location_code                                                
25638063              DOWNSHIRE HILL       LONDON  NW3 1NR   
32905063                  FLASK WALK       LONDON  NW3 1HE   
32907063                  FLASK WALK       LONDON  NW3 1HE   
32907063                  FLASK WALK       LONDON  NW3 1HE   
32912063                  FLASK WALK       LONDON  NW3 1HE   
...                              ...          ...      ...   
12358523000    HAMPSTEAD HIGH STREET       LONDON  NW3 1QD   
12514325000            PERRINS COURT       LONDON  NW3 1QS   
13026783000                  NEW END       LONDON  NW3 1JD   
13174048000    HAMPSTEAD HIGH STREET       LONDON  NW3 1QP   
13174051000    HAMPSTEAD HIGH STREET       LONDON  NW3 1QP   

                                                    use_category  ...  \
location_code                                                     ...   
25638063       Surgeries, Clinics, Health Centres (Rental Val...  ...   
32905063                          Offices (Inc Computer Centres)  ...   
32907063                                                   Shops  ...   
32907063                                                   Shops  ...   
32912063                                                   Shops  ...   
...                                                          ...  ...   
12358523000                       Offices (Inc Computer Centres)  ...   
12514325000                                                Shops  ...   
13026783000                         Showhouses (National Scheme)  ...   
13174048000                                                Shops  ...   
13174051000                       Offices (Inc Computer Centres)  ...   

               duration study-start time-of-entry  time-at-vacant  vacant  \
location_code                                                               
25638063           3524  2010-04-01      494 days            4018       0   
32905063           4018  2010-04-01        0 days            4

In [76]:
subset.to_csv('final-hampstead.csv')

***

In [77]:
## DOUBLE CHECK WE ARE NOT MISSING ANY WHO WERE THERE FROM START TO END (IE. 4019 DAYS)

In [78]:
counter = 0 
stoppingCondition = len(locations)
df2 = pd.DataFrame(data=None, columns=df.columns, index=None)

while counter < stoppingCondition:
        i = locations[counter]
        m = df.loc[i, :].iloc[[ df.loc[i, :]['status_duration'].argmax()]]
        df2 = df2.append(m, ignore_index=False)
        counter += 1

In [79]:
df2

Unnamed: 0       rates_code rates_authority_id rates_authority  \
25638063         25871   00184001B00002          E09000007          Camden   
32905063         24639   00159001009992          E09000007          Camden   
32907063         24727   00159001200006          E09000007          Camden   
32912063         24419   00159000410016          E09000007          Camden   
32945063         24815   00159001500009          E09000007          Camden   
...                ...              ...                ...             ...   
12358523000     399173   00153003210013          E09000007          Camden   
12514325000     397925     TEMP33949894          E09000007          Camden   
13026783000     429519  000760029300001          E09000007          Camden   
13174048000     429164   00153007210028          E09000007          Camden   
13174051000     429169   00153007210029          E09000007          Camden   

                          region                      address_no  \
25638063     Inner London - West                              1B   
32905063     Inner London - West                              10   
32907063     Inner London - West                              12   
32912063     Inner London - West                               4   
32945063     Inner London - West                              15   
...                          ...                             ...   
12358523000  Inner London - West         1ST FLR (SOUTH WEST) 32   
12514325000  Inner London - West  GND FLR 7-9 & 1ST FLR LEFT 7-9   
13026783000  Inner London - West                     APARTMENT 8   
13174048000  Inner London - West                      GND FLR 72   
13174051000  Inner London - West                      1ST FLR 72   

                    address_street address_town  pc__pcs  \
25638063            DOWNSHIRE HILL       LONDON  NW3 1NR   
32905063                FLASK WALK       LONDON  NW3 1HE   
32907063                FLASK WALK       LONDON  NW3 1HE   
32912063                FLASK WALK       LONDON  NW3 1HE   
32945063                FLASK WALK       LONDON  NW3 1HJ   
...                            ...          ...      ...   
12358523000  HAMPSTEAD HIGH STREET       LONDON  NW3 1QD   
12514325000          PERRINS COURT       LONDON  NW3 1QS   
13026783000                NEW END       LONDON  NW3 1JD   
13174048000  HAMPSTEAD HIGH STREET       LONDON  NW3 1QP   
13174051000  HAMPSTEAD HIGH STREET       LONDON  NW3 1QP   

                                                  use_category  ... duration  \
25638063     Surgeries, Clinics, Health Centres (Rental Val...  ...     3524   
32905063                        Offices (Inc Computer Centres)  ...     4018   
32907063                                                 Shops  ...     2192   
32912063                                                 Shops  ...     2722   
32945063                                                 Shops  ...     1461   
...                                                        ...  ...      ...   
12358523000                     Offices (Inc Computer Centres)  ...     1461   
12514325000                                              Shops  ...     1461   
13026783000                       Showhouses (National Scheme)  ...      134   
13174048000                                              Shops  ...      389   
13174051000                     Offices (Inc Computer Centres)  ...      389   

            study-start time-of-entry  time-at-vacant  vacant survival entry  \
25638063     2010-04-01      494 days            4018       0     3524   494   
32905063     2010-04-01        0 days            4018       0     4018     0   
32907063     2010-04-01     1826 days            4018       0     2192  1826   
32912063     2010-04-01     1296 days            4018       0     2722  1296   
32945063     2010-04-01     2557 days            4018       0     1461  2557   
...                 ...           ...             ...     ...      ...   ...   
12358523000  201

In [80]:
df3 = subset.append(df2)

In [81]:
df3

Unnamed: 0       rates_code rates_authority_id rates_authority  \
25638063         25871   00184001B00002          E09000007          Camden   
32905063         24639   00159001009992          E09000007          Camden   
32907063         24701   00159001200006          E09000007          Camden   
32907063         24727   00159001200006          E09000007          Camden   
32912063         24389   00159000410016          E09000007          Camden   
...                ...              ...                ...             ...   
12358523000     399173   00153003210013          E09000007          Camden   
12514325000     397925     TEMP33949894          E09000007          Camden   
13026783000     429519  000760029300001          E09000007          Camden   
13174048000     429164   00153007210028          E09000007          Camden   
13174051000     429169   00153007210029          E09000007          Camden   

                          region                      address_no  \
25638063     Inner London - West                              1B   
32905063     Inner London - West                              10   
32907063     Inner London - West                              12   
32907063     Inner London - West                              12   
32912063     Inner London - West                               4   
...                          ...                             ...   
12358523000  Inner London - West         1ST FLR (SOUTH WEST) 32   
12514325000  Inner London - West  GND FLR 7-9 & 1ST FLR LEFT 7-9   
13026783000  Inner London - West                     APARTMENT 8   
13174048000  Inner London - West                      GND FLR 72   
13174051000  Inner London - West                      1ST FLR 72   

                    address_street address_town  pc__pcs  \
25638063            DOWNSHIRE HILL       LONDON  NW3 1NR   
32905063                FLASK WALK       LONDON  NW3 1HE   
32907063                FLASK WALK       LONDON  NW3 1HE   
32907063                FLASK WALK       LONDON  NW3 1HE   
32912063                FLASK WALK       LONDON  NW3 1HE   
...                            ...          ...      ...   
12358523000  HAMPSTEAD HIGH STREET       LONDON  NW3 1QD   
12514325000          PERRINS COURT       LONDON  NW3 1QS   
13026783000                NEW END       LONDON  NW3 1JD   
13174048000  HAMPSTEAD HIGH STREET       LONDON  NW3 1QP   
13174051000  HAMPSTEAD HIGH STREET       LONDON  NW3 1QP   

                                                  use_category  ... duration  \
25638063     Surgeries, Clinics, Health Centres (Rental Val...  ...     3524   
32905063                        Offices (Inc Computer Centres)  ...     4018   
32907063                                                 Shops  ...     1644   
32907063                                                 Shops  ...     2192   
32912063                                                 Shops  ...     1279   
...                                                        ...  ...      ...   
12358523000                     Offices (Inc Computer Centres)  ...     1461   
12514325000                                              Shops  ...     1461   
13026783000                       Showhouses (National Scheme)  ...      134   
13174048000                                              Shops  ...      389   
13174051000                     Offices (Inc Computer Centres)  ...      389   

            study-start time-of-entry  time-at-vacant  vacant survival entry  \
25638063     2010-04-01      494 days            4018       0     3524   494   
32905063     2010-04-01        0 days            4018       0     4018     0   
32907063     2010-04-01        0 days            1644       0     1644     0   
32907063     2010-04-01     1826 days            4018       0     2192  1826   
32912063     2010-04-01        0 days            1279       0     1279     0   
...                 ...           ...             ...     ...      ...   ...   
12358523000  201

In [82]:
df3 = df3.drop_duplicates()

In [83]:
df3

Unnamed: 0      rates_code rates_authority_id rates_authority  \
25638063         25871  00184001B00002          E09000007          Camden   
32905063         24639  00159001009992          E09000007          Camden   
32907063         24701  00159001200006          E09000007          Camden   
32907063         24727  00159001200006          E09000007          Camden   
32912063         24389  00159000410016          E09000007          Camden   
...                ...             ...                ...             ...   
7993663000      319145  00160000600007          E09000007          Camden   
8773536000      357261  99121006800016          E09000007          Camden   
11040907000     386828  99121001500033          E09000007          Camden   
11040913000     386804  99121001500022          E09000007          Camden   
11765732000     419532  0029500482000A          E09000007          Camden   

                          region        address_no   address_street  \
25638063     Inner London - West                1B   DOWNSHIRE HILL   
32905063     Inner London - West                10       FLASK WALK   
32907063     Inner London - West                12       FLASK WALK   
32907063     Inner London - West                12       FLASK WALK   
32912063     Inner London - West                 4       FLASK WALK   
...                          ...               ...              ...   
7993663000   Inner London - West         STORES AT  STREATLEY PLACE   
8773536000   Inner London - West  BST & GRD FLR 68     HEATH STREET   
11040907000  Inner London - West         BST 15-17     HEATH STREET   
11040913000  Inner London - West     GND FLR 15-17     HEATH STREET   
11765732000  Inner London - West                48     ROSSLYN HILL   

            address_town  pc__pcs  \
25638063          LONDON  NW3 1NR   
32905063          LONDON  NW3 1HE   
32907063          LONDON  NW3 1HE   
32907063          LONDON  NW3 1HE   
32912063          LONDON  NW3 1HE   
...                  ...      ...   
7993663000        LONDON  NW3 1HP   
8773536000        LONDON  NW3 1DN   
11040907000       LONDON  NW3 6TR   
11040913000       LONDON  NW3 6TR   
11765732000       LONDON  NW3 1NH   

                                                  use_category  ... duration  \
25638063     Surgeries, Clinics, Health Centres (Rental Val...  ...     3524   
32905063                        Offices (Inc Computer Centres)  ...     4018   
32907063                                                 Shops  ...     1644   
32907063                                                 Shops  ...     2192   
32912063                                                 Shops  ...     1279   
...                                                        ...  ...      ...   
7993663000                                              Stores  ...     2557   
8773536000                                               Shops  ...      921   
11040907000                     Offices (Inc Computer Centres)  ...     2192   
11040913000                                              Shops  ...     2192   
11765732000                                        Restaurants  ...      625   

            study-start time-of-entry  time-at-vacant  vacant survival entry  \
25638063     2010-04-01      494 days            4018       0     3524   494   
32905063     2010-04-01        0 days            4018       0     4018     0   
32907063     2010-04-01        0 days            1644       0     1644     0   
32907063     2010-04-01     1826 days            4018       0     2192  1826   
32912063     2010-04-01        0 days            1279       0     1279     0   
...                 ...           ...             ...     ...      ...   ...   
7993663000   2010-04-01     1461 days            4018       0     2557  1461   
8773536000   2010-04-01     1454 days            2375       0      921  1454   
11040907000  2010-04-01     1826 days            4018       0     2192  1826   
11040913000  2010-04-01     

In [363]:
## if time-at-vacant = 4018 then Vacant change to 1 (ie it died)

In [85]:
check1 = df3.loc[df3['time-at-vacant'] < 4018]

In [86]:
check1

Unnamed: 0      rates_code rates_authority_id rates_authority  \
32907063        24701  00159001200006          E09000007          Camden   
32912063        24389  00159000410016          E09000007          Camden   
32948063        24439  0015900053001B          E09000007          Camden   
41419063        23156  00153007100001          E09000007          Camden   
41422063        23330  0015300751003B          E09000007          Camden   
...               ...             ...                ...             ...   
111990063       21595  001530006A0053          E09000007          Camden   
194153063       21551  001530006A0034          E09000007          Camden   
272838000      338860  0006900821001B          E09000007          Camden   
6477509000      23441  00153007910012          E09000007          Camden   
8773536000     357261  99121006800016          E09000007          Camden   

                         region        address_no         address_street  \
32907063    Inner London - West                12             FLASK WALK   
32912063    Inner London - West                 4             FLASK WALK   
32948063    Inner London - West                 5             FLASK WALK   
41419063    Inner London - West                71  HAMPSTEAD HIGH STREET   
41422063    Inner London - West                75  HAMPSTEAD HIGH STREET   
...                         ...               ...                    ...   
111990063   Inner London - West     1ST F REAR 6A  HAMPSTEAD HIGH STREET   
194153063   Inner London - West    1ST F FRONT 6A  HAMPSTEAD HIGH STREET   
272838000   Inner London - West   BST & GND FS 82           HEATH STREET   
6477509000  Inner London - West               79B  HAMPSTEAD HIGH STREET   
8773536000  Inner London - West  BST & GRD FLR 68           HEATH STREET   

           address_town  pc__pcs                    use_category  ...  \
32907063         LONDON  NW3 1HE                           Shops  ...   
32912063         LONDON  NW3 1HE                           Shops  ...   
32948063         LONDON  NW3 1HJ                           Shops  ...   
41419063         LONDON  NW3 1QP                     Restaurants  ...   
41422063         LONDON  NW3 1QX                           Shops  ...   
...                 ...      ...                             ...  ...   
111990063        LONDON  NW3 1PR  Offices (Inc Computer Centres)  ...   
194153063        LONDON  NW3 1PR  Offices (Inc Computer Centres)  ...   
272838000        LONDON  NW3 1DN                           Shops  ...   
6477509000       LONDON  NW3 1RE                           Shops  ...   
8773536000       LONDON  NW3 1DN                           Shops  ...   

           duration study-start time-of-entry  time-at-vacant  vacant  \
32907063       1644  2010-04-01        0 days            1644       0   
32912063       1279  2010-04-01        0 days            1279       0   
32948063       1826  2010-04-01        0 days            1826       0   
41419063       1187  2010-04-01        0 days            1187       0   
41422063       1006  2010-04-01        0 days            1006       0   
...             ...         ...           ...             ...     ...   
111990063      3287  2010-04-01        0 days            3287       0   
194153063      3287  2010-04-01        0 days            3287       0   
272838000      1834  2010-04-01      267 days            2101       0   
6477509000     1478  2010-04-01     1627 days            3105       0   
8773536000      921  2010-04-01     1454 days            2375       0   

           survival entry stat  StatusChange TenantChange  
32907063       1644     0    1         False         True  
32912063       1279     0    1         False         True  
32948063       1826     0    1         False         True  
41419063       1187     0    1         False         True  
41422063       1006     0    1         False         True  
...             ...   ...  ...           ...          ...  
111990063  

In [87]:
check2 = check1.drop_duplicates()

In [88]:
check2

Unnamed: 0      rates_code rates_authority_id rates_authority  \
32907063        24701  00159001200006          E09000007          Camden   
32912063        24389  00159000410016          E09000007          Camden   
32948063        24439  0015900053001B          E09000007          Camden   
41419063        23156  00153007100001          E09000007          Camden   
41422063        23330  0015300751003B          E09000007          Camden   
...               ...             ...                ...             ...   
111990063       21595  001530006A0053          E09000007          Camden   
194153063       21551  001530006A0034          E09000007          Camden   
272838000      338860  0006900821001B          E09000007          Camden   
6477509000      23441  00153007910012          E09000007          Camden   
8773536000     357261  99121006800016          E09000007          Camden   

                         region        address_no         address_street  \
32907063    Inner London - West                12             FLASK WALK   
32912063    Inner London - West                 4             FLASK WALK   
32948063    Inner London - West                 5             FLASK WALK   
41419063    Inner London - West                71  HAMPSTEAD HIGH STREET   
41422063    Inner London - West                75  HAMPSTEAD HIGH STREET   
...                         ...               ...                    ...   
111990063   Inner London - West     1ST F REAR 6A  HAMPSTEAD HIGH STREET   
194153063   Inner London - West    1ST F FRONT 6A  HAMPSTEAD HIGH STREET   
272838000   Inner London - West   BST & GND FS 82           HEATH STREET   
6477509000  Inner London - West               79B  HAMPSTEAD HIGH STREET   
8773536000  Inner London - West  BST & GRD FLR 68           HEATH STREET   

           address_town  pc__pcs                    use_category  ...  \
32907063         LONDON  NW3 1HE                           Shops  ...   
32912063         LONDON  NW3 1HE                           Shops  ...   
32948063         LONDON  NW3 1HJ                           Shops  ...   
41419063         LONDON  NW3 1QP                     Restaurants  ...   
41422063         LONDON  NW3 1QX                           Shops  ...   
...                 ...      ...                             ...  ...   
111990063        LONDON  NW3 1PR  Offices (Inc Computer Centres)  ...   
194153063        LONDON  NW3 1PR  Offices (Inc Computer Centres)  ...   
272838000        LONDON  NW3 1DN                           Shops  ...   
6477509000       LONDON  NW3 1RE                           Shops  ...   
8773536000       LONDON  NW3 1DN                           Shops  ...   

           duration study-start time-of-entry  time-at-vacant  vacant  \
32907063       1644  2010-04-01        0 days            1644       0   
32912063       1279  2010-04-01        0 days            1279       0   
32948063       1826  2010-04-01        0 days            1826       0   
41419063       1187  2010-04-01        0 days            1187       0   
41422063       1006  2010-04-01        0 days            1006       0   
...             ...         ...           ...             ...     ...   
111990063      3287  2010-04-01        0 days            3287       0   
194153063      3287  2010-04-01        0 days            3287       0   
272838000      1834  2010-04-01      267 days            2101       0   
6477509000     1478  2010-04-01     1627 days            3105       0   
8773536000      921  2010-04-01     1454 days            2375       0   

           survival entry stat  StatusChange TenantChange  
32907063       1644     0    1         False         True  
32912063       1279     0    1         False         True  
32948063       1826     0    1         False         True  
41419063       1187     0    1         False         True  
41422063       1006     0    1         False         True  
...             ...   ...  ...           ...          ...  
111990063  

In [364]:
df3.loc[df3['time-at-vacant'] < 4018, ['vacant']] = 1

/opt/conda/envs/sds2020/lib/python3.7/site-packages/pandas/core/indexing.py:1765: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  isetter(loc, value)


In [365]:
df3

Unnamed: 0      rates_code rates_authority_id rates_authority  \
25638063         25871  00184001B00002          E09000007          Camden   
32905063         24639  00159001009992          E09000007          Camden   
32907063         24701  00159001200006          E09000007          Camden   
32907063         24727  00159001200006          E09000007          Camden   
32912063         24389  00159000410016          E09000007          Camden   
...                ...             ...                ...             ...   
7993663000      319145  00160000600007          E09000007          Camden   
8773536000      357261  99121006800016          E09000007          Camden   
11040907000     386828  99121001500033          E09000007          Camden   
11040913000     386804  99121001500022          E09000007          Camden   
11765732000     419532  0029500482000A          E09000007          Camden   

                          region        address_no   address_street  \
25638063     Inner London - West                1B   DOWNSHIRE HILL   
32905063     Inner London - West                10       FLASK WALK   
32907063     Inner London - West                12       FLASK WALK   
32907063     Inner London - West                12       FLASK WALK   
32912063     Inner London - West                 4       FLASK WALK   
...                          ...               ...              ...   
7993663000   Inner London - West         STORES AT  STREATLEY PLACE   
8773536000   Inner London - West  BST & GRD FLR 68     HEATH STREET   
11040907000  Inner London - West         BST 15-17     HEATH STREET   
11040913000  Inner London - West     GND FLR 15-17     HEATH STREET   
11765732000  Inner London - West                48     ROSSLYN HILL   

            address_town  pc__pcs  \
25638063          LONDON  NW3 1NR   
32905063          LONDON  NW3 1HE   
32907063          LONDON  NW3 1HE   
32907063          LONDON  NW3 1HE   
32912063          LONDON  NW3 1HE   
...                  ...      ...   
7993663000        LONDON  NW3 1HP   
8773536000        LONDON  NW3 1DN   
11040907000       LONDON  NW3 6TR   
11040913000       LONDON  NW3 6TR   
11765732000       LONDON  NW3 1NH   

                                                  use_category  ... duration  \
25638063     Surgeries, Clinics, Health Centres (Rental Val...  ...     3524   
32905063                        Offices (Inc Computer Centres)  ...     4018   
32907063                                                 Shops  ...     1644   
32907063                                                 Shops  ...     2192   
32912063                                                 Shops  ...     1279   
...                                                        ...  ...      ...   
7993663000                                              Stores  ...     2557   
8773536000                                               Shops  ...      921   
11040907000                     Offices (Inc Computer Centres)  ...     2192   
11040913000                                              Shops  ...     2192   
11765732000                                        Restaurants  ...      625   

            study-start time-of-entry  time-at-vacant  vacant survival entry  \
25638063     2010-04-01      494 days            4018       0     3524   494   
32905063     2010-04-01        0 days            4018       0     4018     0   
32907063     2010-04-01        0 days            1644       1     1644     0   
32907063     2010-04-01     1826 days            4018       0     2192  1826   
32912063     2010-04-01        0 days            1279       1     1279     0   
...                 ...           ...             ...     ...      ...   ...   
7993663000   2010-04-01     1461 days            4018       0     2557  1461   
8773536000   2010-04-01     1454 days            2375       1      921  1454   
11040907000  2010-04-01     1826 days            4018       0     2192  1826   
11040913000  2010-04-01     

In [366]:
df3.to_csv('final.csv')